From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shawn Hoffman <business@symbrkrs.com>
Date: Sun, 10 May 2026 21:51:36 -0700
Subject: [PATCH] 9679: nds32 bug fixes

nds32 bug fixes: * sp_bound_priv,sp_base,sp_base_priv: fix
 csreg addr. * fix data endianness issues. * handle mfusr/mtusr PC correctly.
 * scw: inhibit decompiler optimizing out check. * tlbop Rt,Ra,Probe: indicate
 Rt is written. * dpref: fix shift amount. * cctl: indicate Rt may be written.
 * standby: allow fallthrough. * addi.gp: display Rt, Imm19s in disasm. *
 *_srli: fix shift direction. * bgezal/bltzal/jralnez/jrnez: fix branch
 polarity. * abs: fix pcode (implement saturation behavior). * ave: fix shift
 amount and don't chop off high bits. * clz/clo: use lzcount pcodeop instead
 of loop. * add_abs_diff: fix intermediate truncation. * add5.pc: fix addr
 computation. * bmski33: fix pcode. * pop25: fix sp calculations (pop imm5
 first; use Lmwai instead of LmwOp). * FP ops: fix pcodeops being called;
 indicate that Fxt is input AND output. * fmtcsr: constrain to Fsa=0.

added instructions:
* mtlbi/mtlei: just set lb/le registers.
* ifcall/ifret: unimpl for now.

misc:
* push25: handle setting s2 without branch.
* add branch_if/call_if macros. should help with IFCALL support.

note:
The fixes to abs and add_abs_diff make the pcode correct, however the
decompiler output may not be ideal.
---
 .../Processors/NDS32/data/languages/lsmw.sinc |   5 +
 .../NDS32/data/languages/nds32.pspec          |   6 +-
 .../NDS32/data/languages/nds32.sinc           | 280 +++++++++++-------
 3 files changed, 174 insertions(+), 117 deletions(-)

diff --git a/Ghidra/Processors/NDS32/data/languages/lsmw.sinc b/Ghidra/Processors/NDS32/data/languages/lsmw.sinc
index bff5e75888..ba307bee67 100644
--- a/Ghidra/Processors/NDS32/data/languages/lsmw.sinc
+++ b/Ghidra/Processors/NDS32/data/languages/lsmw.sinc
@@ -31,6 +31,11 @@ Dreg: gp is gp & regNum=29 { export gp; }
 Dreg: lp is lp & regNum=30 { export lp; }
 Dreg: sp is sp & regNum=31 { export sp; }
 
+macro Lmwai(reg) {
+	reg = *mult_addr;
+	mult_addr = mult_addr + 4;
+}
+
 macro Smwad(reg) {
     mult_addr = mult_addr - 4;
     *mult_addr = reg;
diff --git a/Ghidra/Processors/NDS32/data/languages/nds32.pspec b/Ghidra/Processors/NDS32/data/languages/nds32.pspec
index d70bdf4abd..3d1b1679b1 100644
--- a/Ghidra/Processors/NDS32/data/languages/nds32.pspec
+++ b/Ghidra/Processors/NDS32/data/languages/nds32.pspec
@@ -128,9 +128,9 @@
          <symbol name="fucpr"          address="csreg:0x228"      size="4" description="" />
          <symbol name="hsp_ctl"        address="csreg:0x230"      size="4" description="" />
          <symbol name="sp_bound"       address="csreg:0x231"      size="4" description="" />
-         <symbol name="sp_bound_priv"  address="csreg:0x230"      size="4" description="" />
-         <symbol name="sp_base"        address="csreg:0x230"      size="4" description="" />
-         <symbol name="sp_base_priv"   address="csreg:0x230"      size="4" description="" />
+         <symbol name="sp_bound_priv"  address="csreg:0x232"      size="4" description="" />
+         <symbol name="sp_base"        address="csreg:0x233"      size="4" description="" />
+         <symbol name="sp_base_priv"   address="csreg:0x234"      size="4" description="" />
 
          <symbol name="dma_cfg"        address="csreg:0x280"      size="4" description="" />
          <symbol name="dma_gcsw"       address="csreg:0x288"      size="4" description="" />
diff --git a/Ghidra/Processors/NDS32/data/languages/nds32.sinc b/Ghidra/Processors/NDS32/data/languages/nds32.sinc
index eb6f8c8fdf..105ce647e4 100644
--- a/Ghidra/Processors/NDS32/data/languages/nds32.sinc
+++ b/Ghidra/Processors/NDS32/data/languages/nds32.sinc
@@ -1,6 +1,8 @@
 ### General ###
 
-define endian=big;
+# Instruction endianness is always big, but data load/store endian is controlled by PSW.BE
+# We support staticly setting data endianness (not dynamically via PSW.BE)
+define endian=$(ENDIAN);
 define alignment=2;
 define space ram type=ram_space size=4 wordsize=1 default;
 define space register type=register_space size=4;
@@ -18,8 +20,13 @@ define register offset=0x90 size=4
 define register offset=0x100 size=8
 [d0 d1];
 
+@if ENDIAN == "little"
+define register offset=0x100 size=4
+[d0.lo d0.hi d1.lo d1.hi];
+@else
 define register offset=0x100 size=4
 [d0.hi d0.lo d1.hi d1.lo];
+@endif
 
 define register offset=0x200 size=4
 [ itb lb lc le ifc_lp
@@ -51,7 +58,7 @@ define context contextreg
 	regNum		= (27,31)	# register for load/store multiple instructions
 ;
 
-define token instr32(32)
+define token instr32(32) endian=big
     OpSz        = (31, 31)
     Opc         = (25, 30)
     Rt          = (20, 24)
@@ -222,6 +229,16 @@ attach variables [ Fdt Fda Fdb ]
 @define GPR     "(Alu2Mod=0b0001)"
 
 
+macro branch_if(cond, target) {
+    if (cond) goto target;
+}
+
+macro call_if(cond, target) {
+    if (!cond) goto inst_next;
+    call target;
+}
+
+
 ### ALU Instruction with Immediate ###
 
 :addi  Rt, Ra, Imm15s is $(I32) & Opc=0b101000 & Rt & Ra & Imm15s { Rt = Ra + Imm15s; }
@@ -321,9 +338,10 @@ UsrName: "pfm_ctl"     is Group=2 & Usr=4 { tmp:2 = 0x208; export *[csreg]:4 tmp
 
 
 :mfusr Rt, UsrName is $(I32) & $(ALU_2) & Rt & UsrName & $(ALU2Z) & Sub6=0b100000 { Rt = UsrName; }
-:mfusr Rt, pc is $(I32) & $(ALU_2) & Rt & Group=0 & Usr=0b11111 & $(ALU2Z) & Sub6=0b100000  & pc { Rt = inst_next; } 
+:mfusr Rt, pc is $(I32) & $(ALU_2) & Rt & Group=0 & Usr=0b11111 & $(ALU2Z) & Sub6=0b100000 & pc { Rt = inst_start; }
 :mtusr Rt, UsrName is $(I32) & $(ALU_2) & Rt & UsrName & $(ALU2Z) & Sub6=0b100001 { UsrName = Rt; }
-:mtusr Rt, pc is $(I32) & $(ALU_2) & Rt & Group=0 & Usr=0b11111 & $(ALU2Z) & Sub6=0b100001 & pc { pc = Rt; goto[pc]; } # Not sure this works correctly
+# The manual explicitly states pc cannot be written.
+:mtusr Rt, pc is $(I32) & $(ALU_2) & Rt & Group=0 & Usr=0b11111 & $(ALU2Z) & Sub6=0b100001 & pc unimpl
 
 
 ### Divide Instructions ###
@@ -422,8 +440,17 @@ LsmwM_: "m" is LsmwRa & LsmwM=1 { LsmwRa = mult_addr; }
 
 ### Load / Store Instruction for Atomic Updates ###
 
+define pcodeop scw_succeed;
+
 :llw Rt, AddrRaRbsv is $(I32) & $(MEM) & Rt & AddrRaRbsv & Sub8=0b00011000 { Rt = *AddrRaRbsv; }
-:scw Rt, AddrRaRbsv is $(I32) & $(MEM) & Rt & AddrRaRbsv & Sub8=0b00011001 { *AddrRaRbsv = Rt; }
+:scw Rt, AddrRaRbsv is $(I32) & $(MEM) & Rt & AddrRaRbsv & Sub8=0b00011001 {
+    # scw_succeed is used to inhibit decompiler optimizing out the check
+    local ok:4 = scw_succeed(AddrRaRbsv:4, Rt:4);
+    if (ok == 0) goto <fail>;
+        *AddrRaRbsv = Rt;
+    <fail>
+    Rt = ok;
+}
 
 
 ### Load / Store Instructions with User-mode Privilege ###
@@ -449,32 +476,28 @@ Rel24: addr is Imm24s [ addr = inst_start + (Imm24s << 1); ] { export *:4 addr;
 Rel14: addr is Imm14s [ addr = inst_start + (Imm14s << 1); ] { export *:4 addr; }
 Rel16: addr is Imm16s [ addr = inst_start + (Imm16s << 1); ] { export *:4 addr; }
 
-:beq  Rt, Ra, Rel14 is $(I32) & $(BR1) & Rt & Ra & Br1t=0 & Rel14 { if(Rt == Ra) goto Rel14; }
-:bne  Rt, Ra, Rel14 is $(I32) & $(BR1) & Rt & Ra & Br1t=1 & Rel14 { if(Rt != Ra) goto Rel14; }
-:beqz Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0010 & Rel16 { if(Rt == 0)  goto Rel16; }
-:bnez Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0011 & Rel16 { if(Rt != 0)  goto Rel16; }
-:bgez Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0100 & Rel16 { if(Rt s>= 0) goto Rel16; }
-:bltz Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0101 & Rel16 { if(Rt s< 0)  goto Rel16; }
-:bgtz Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0110 & Rel16 { if(Rt s> 0)  goto Rel16; }
-:blez Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0111 & Rel16 { if(Rt s<= 0) goto Rel16; }
+:beq  Rt, Ra, Rel14 is $(I32) & $(BR1) & Rt & Ra & Br1t=0 & Rel14 { branch_if(Rt == Ra, Rel14); }
+:bne  Rt, Ra, Rel14 is $(I32) & $(BR1) & Rt & Ra & Br1t=1 & Rel14 { branch_if(Rt != Ra, Rel14); }
+:beqz Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0010 & Rel16 { branch_if(Rt == 0,  Rel16); }
+:bnez Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0011 & Rel16 { branch_if(Rt != 0,  Rel16); }
+:bgez Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0100 & Rel16 { branch_if(Rt s>= 0, Rel16); }
+:bltz Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0101 & Rel16 { branch_if(Rt s< 0,  Rel16); }
+:bgtz Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0110 & Rel16 { branch_if(Rt s> 0,  Rel16); }
+:blez Rt,     Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b0111 & Rel16 { branch_if(Rt s<= 0, Rel16); }
 
 
 ### Branch with link Instruction ###
 
 :bgezal Rt, Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b1100 & Rel16
-{   
+{
     lp = inst_next;
-    if(Rt s>= 0) goto <end>;
-        call Rel16;
-    <end>
+    call_if(Rt s>= 0, Rel16);
 }
 
 :bltzal Rt, Rel16 is $(I32) & $(BR2) & Rt & Br2t=0b1101 & Rel16
-{   
+{
     lp = inst_next;
-    if(Rt s< 0) goto <end>;
-        call Rel16;
-    <end>
+    call_if(Rt s< 0, Rel16);
 }
 
 
@@ -516,7 +539,7 @@ define pcodeop TLB_FlushAll;
 :tlbop Ra,"RWrite"      is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=2 & Sub5=0b01110 { TLB_RWrite(Ra:4); }
 :tlbop Ra,"RWriteLock"  is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=3 & Sub5=0b01110 { TLB_RWriteLock(Ra:4); }
 :tlbop Ra,"Unlock"      is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=4 & Sub5=0b01110 { TLB_Unlock(Ra:4); }
-:tlbop Rt,Ra,"Probe"    is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=5 & Sub5=0b01110 { TLB_Probe(Rt:4, Ra:4); }
+:tlbop Rt,Ra,"Probe"    is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=5 & Sub5=0b01110 { Rt = TLB_Probe(Ra:4); }
 :tlbop Ra,"Invalidate"  is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=6 & Sub5=0b01110 { TLB_Invalidate(Ra:4); }
 :tlbop "FlushAll"       is $(I32) & $(MISC) & Rt & Ra & Rb=0 & TlbopSub=7 & Sub5=0b01110 { TLB_FlushAll(); }
 
@@ -552,11 +575,8 @@ define pcodeop isync;
 
 define pcodeop dpref;
 
-OffsetRbsv2: (Rb "<<" sv) is Rb & sv { off:4 = Rb << (sv + 1); export off; }
-AddrRaRbsv2: [Ra + OffsetRbsv2] is Ra & OffsetRbsv2 { addr:4 = Ra + OffsetRbsv2; export addr; }
-
-:dpref DprefSub, AddrRaRbsv2 is $(I32) & $(MEM) & DprefD=0 & DprefSub & AddrRaRbsv2 & Sub8=0b00010011 {
-    dpref(DprefSub:1, AddrRaRbsv2:4);
+:dpref DprefSub, AddrRaRbsv is $(I32) & $(MEM) & DprefD=0 & DprefSub & AddrRaRbsv & Sub8=0b00010011 {
+    dpref(DprefSub:1, AddrRaRbsv:4);
 }
 
 DprefD_: "w" is DprefD=0 { }
@@ -623,7 +643,7 @@ define pcodeop trap;
 # TODO : special function, with subfunctions
 define pcodeop cctl;
 
-:cctl Rt, Ra, CctlLevel, CctlSub is $(I32) & $(MISC) & Rt & Ra & CctlZ=0 & CctlLevel & CctlSub & Sub5=0b00001 { cctl(Rt:4, Ra:4, CctlLevel:1, CctlSub:1); }
+:cctl Rt, Ra, CctlLevel, CctlSub is $(I32) & $(MISC) & Rt & Ra & CctlZ=0 & CctlLevel & CctlSub & Sub5=0b00001 { Rt = cctl(Rt:4, Ra:4, CctlLevel:1, CctlSub:1); }
 
 
 # Miscellaneous Instructions (Baseline)
@@ -644,7 +664,8 @@ SetendBE: "b" is Toggle=1 { setend(1:1); }
 
 :setend.^SetendBE is $(I32) & $(MISC) & ToggleL=0 & SetendBE & SrIdx=0b0010000000 & Rd=0b00001 & Sub5=0b00011 { }
 
-:standby StandbySub is $(I32) & $(MISC) & Rt=0 & Ra=0 & Rb=0 & StandbyZ=0 & StandbySub & Sub5=0b00000 { goto inst_start; }
+define pcodeop standby;
+:standby StandbySub is $(I32) & $(MISC) & Rt=0 & Ra=0 & Rb=0 & StandbyZ=0 & StandbySub & Sub5=0b00000 { standby(StandbySub:1); }
 
 
 
@@ -652,7 +673,7 @@ SetendBE: "b" is Toggle=1 { setend(1:1); }
 
 ### ALU Instructions ###
 
-:addi.gp is $(I32) & $(SBGP) & Rt & GpSub1=0b1 & Imm19s { Rt = gp + Imm19s; }
+:addi.gp Rt, Imm19s is $(I32) & $(SBGP) & Rt & GpSub1=0b1 & Imm19s { Rt = gp + Imm19s; }
 
 
 ### Multiply and Divide Instructions (V2) ###
@@ -660,16 +681,15 @@ SetendBE: "b" is Toggle=1 { setend(1:1); }
 :mulr64 Rt, Ra, Rb is $(I32) & $(ALU_2) & Rt & Ra & Rb & $(GPR) & Sub6=0b101001 & Rtl & Rth
 {
     res:8 = zext(Ra) * zext(Rb);
-    Rtl = res[32,32];
-    Rth = res[0,32];
-
+    Rtl = res[0,32];
+    Rth = res[32,32];
 }
 
 :mulsr64 Rt, Ra, Rb is $(I32) & $(ALU_2) & Rt & Ra & Rb & $(GPR) & Sub6=0b101000 & Rtl & Rth
 {
     res:8 = sext(Ra) * sext(Rb);
-    Rtl = res[32,32];
-    Rth = res[0,32];
+    Rtl = res[0,32];
+    Rth = res[32,32];
 }
 
 :maddr32 Rt, Ra, Rb is $(I32) & $(ALU_2) & Rt & Ra & Rb & $(GPR) & Sub6=0b110011 { Rt = Rt + (Ra * Rb); }
@@ -723,20 +743,29 @@ GpWordAddress: [+ off] is Imm17s [ off = Imm17s << 2; ] { addr:4 = gp + off; exp
 :sub_slli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b00001 { Rt = Ra - (Rb << sh); }
 :xor_slli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b00011 { Rt = Ra ^ (Rb << sh); }
 
-:add_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11100 { Rt = Ra + (Rb << sh); }
-:and_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11110 { Rt = Ra & (Rb << sh); }
-:or_srli  Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b10101 { Rt = Ra | (Rb << sh); }
-:sub_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11101 { Rt = Ra - (Rb << sh); }
-:xor_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11111 { Rt = Ra ^ (Rb << sh); }
+:add_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11100 { Rt = Ra + (Rb >> sh); }
+:and_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11110 { Rt = Ra & (Rb >> sh); }
+:or_srli  Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b10101 { Rt = Ra | (Rb >> sh); }
+:sub_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11101 { Rt = Ra - (Rb >> sh); }
+:xor_srli Rt, Ra, Rb, sh is $(I32) & $(ALU_1) & Rt & Ra & Rb & sh & Sub5=0b11111 { Rt = Ra ^ (Rb >> sh); }
 
 ### Conditional Branch and Jump Instructions (V3) ###
 
 Rel8: addr is Imm8s [ addr = inst_start + (Imm8s << 1); ] { export *:4 addr; }
-:beqc  Rt, Imm11s, Rel8 is $(I32) & $(BR3) & Rt & Bxxc=0 & Imm11s & Rel8 { if(Rt == Imm11s) goto Rel8; }
-:bnec  Rt, Imm11s, Rel8 is $(I32) & $(BR3) & Rt & Bxxc=1 & Imm11s & Rel8 { if(Rt != Imm11s) goto Rel8; }
+:beqc  Rt, Imm11s, Rel8 is $(I32) & $(BR3) & Rt & Bxxc=0 & Imm11s & Rel8 { branch_if(Rt == Imm11s, Rel8); }
+:bnec  Rt, Imm11s, Rel8 is $(I32) & $(BR3) & Rt & Bxxc=1 & Imm11s & Rel8 { branch_if(Rt != Imm11s, Rel8); }
 
-:jralnez Rt,Rb is $(I32) & $(JREG) & Rt & Ra=0 & Rb & DtIt=0b00 & Jz=0 & JrHint=0 & Sub5=0b00011 { if(Rb == 0) goto <end>; Rt = inst_next; call [Rb]; <end> }
-:jrnez   Rb    is $(I32) & $(JREG) & Rt=0 & Ra=0 & Rb & DtIt=0b00 & Jz=0 & JrHint=0 & Sub5=0b00010 { if(Rb == 0) goto <end>; goto [Rb]; <end> }
+:jralnez Rt,Rb is $(I32) & $(JREG) & Rt & Ra=0 & Rb & DtIt=0b00 & Jz=0 & JrHint=0 & Sub5=0b00011 {
+    Rt = inst_next;
+    if(Rb == 0) goto <end>;
+        call [Rb];
+    <end>
+}
+:jrnez   Rb    is $(I32) & $(JREG) & Rt=0 & Ra=0 & Rb & DtIt=0b00 & Jz=0 & JrHint=0 & Sub5=0b00010 {
+    if(Rb == 0) goto <end>;
+        goto [Rb];
+    <end>
+}
 
 ### Bit Manipulation Instructions (V3) ###
 
@@ -755,14 +784,17 @@ Rel8: addr is Imm8s [ addr = inst_start + (Imm8s << 1); ] { export *:4 addr; }
 
 :abs Rt, Ra is $(I32) & $(ALU_2) & Rt & Ra & Rb=0 & $(ALU2Z) & Sub6=0b000011
 {
-    gez:4 = zext(Ra s>= 0);
-    ltz:4 = zext(Ra s< 0);
-    Rt = (Ra * gez) | ((-Ra) * ltz);
+    local sign:4 = Ra s>> 31;
+    local mag:4 = (Ra ^ sign) - sign;
+    local sat:4 = mag s>> 31;
+    Rt = mag ^ sat;
 }
 
 :ave Rt, Ra, Rb is $(I32) & $(ALU_2) & Rt & Ra & Rb & $(ALU2Z) & Sub6=0b000010
 {
-    Rt = (Ra + Rb + 1) s>> 2;
+    local sum:8 = sext(Ra) + sext(Rb) + 1;
+    local shifted:8 = sum s>> 1;
+    Rt = shifted:4;
 }
 
 :max Rt, Ra, Rb is $(I32) & $(ALU_2) & Rt & Ra & Rb & $(ALU2Z) & Sub6=0b000000
@@ -816,34 +848,12 @@ Rel8: addr is Imm8s [ addr = inst_start + (Imm8s << 1); ] { export *:4 addr; }
 
 :clz Rt, Ra is $(I32) & $(ALU_2) & Rt & Ra & Imm5u=0 & $(ALU2Z) & Sub6=0b000111
 {
-  countTmp:4 = 0;
-  inputTmp:4 = Ra;
-
- <loopbegin>
-  if ((inputTmp & 0x80000000) != 0) goto <loopend>;
-
-  countTmp = countTmp + 1;
-  inputTmp = (inputTmp << 1) | 1;
-  goto <loopbegin>;
-
- <loopend>
-  Rt = countTmp;
+  Rt = lzcount(Ra);
 }
 
 :clo Rt, Ra is $(I32) & $(ALU_2) & Rt & Ra & Imm5u=0 & $(ALU2Z) & Sub6=0b000110
 {
-  countTmp:4 = 0;
-  inputTmp:4 = Ra;
-
- <loopbegin>
-  if ((inputTmp & 0x80000000) == 0) goto <loopend>;
-
-  countTmp = countTmp + 1;
-  inputTmp = (inputTmp << 1) | 1;
-  goto <loopbegin>;
-
- <loopend>
-  Rt = countTmp;
+  Rt = lzcount(~Ra);
 }
 
 
@@ -856,15 +866,15 @@ Rel8: addr is Imm8s [ addr = inst_start + (Imm8s << 1); ] { export *:4 addr; }
 
 macro add_abs_diff(dst, src1, src2, shift)
 {
-    local src1_ = src1 >> shift;
-    local src2_ = src2 >> shift;
-    local src1__ = src1_:1;
-    local src2__ = src2_:1;
-    local a:1 = src1__ - src2__;
-    local agez:1 = zext(a s>= 0);
-    local altz:1 = zext(a s< 0);
-    local aabs:1 = (a * agez) | ((-a) * altz);
-    dst = dst + zext(aabs);
+    local aa = src1 >> shift;
+    local bb = src2 >> shift;
+    local b1:2 = zext(aa:1);
+    local b2:2 = zext(bb:1);
+    local d:2  = b1 - b2;
+    local sign:2 = d s>> 15;
+    local diff:2 = (d ^ sign) - sign;
+    local abs:1 = diff:1;
+    dst = dst + zext(abs);
 }
 :pbsad Rt, Ra, Rb is $(I32) & $(SIMD) & Rt & Ra & Rb & Rd=0 & Sub5=0b0000
 {
@@ -890,6 +900,7 @@ macro add_abs_diff(dst, src1, src2, shift)
 	m2:1 = (Ra[8,8]  == match);
 	m3:1 = (Ra[16,8] == match);
 	m4:1 = (Ra[24,8] == match);
+@if ENDIAN == "little"
 	Rt = -4;
 	if (m1) goto inst_next;
 	Rt = -3;
@@ -898,9 +909,17 @@ macro add_abs_diff(dst, src1, src2, shift)
 	if (m3) goto inst_next;
 	Rt = -1;
 	if (m4) goto inst_next;
+@else
+	Rt = -4;
+	if (m4) goto inst_next;
+	Rt = -3;
+	if (m3) goto inst_next;
+	Rt = -2;
+	if (m2) goto inst_next;
+	Rt = -1;
+	if (m1) goto inst_next;
+@endif
 	Rt = 0;
-	# choosery method
-	# rd = 0 + (zext(m1)*-4) + (zext(m2)*-3) + (zext(m3)*-2) + (zext(m4)*-1);
 }
 
 :ffbi Rt, Ra, Imm8u is $(I32) & $(ALU_2) & Rt & Ra & Imm8u & Sub7=0b1001110 {
@@ -909,6 +928,7 @@ macro add_abs_diff(dst, src1, src2, shift)
 	m2:1 = (Ra[8,8]  == match);
 	m3:1 = (Ra[16,8] == match);
 	m4:1 = (Ra[24,8] == match);
+@if ENDIAN == "little"
 	Rt = -4;
 	if (m1) goto inst_next;
 	Rt = -3;
@@ -917,6 +937,16 @@ macro add_abs_diff(dst, src1, src2, shift)
 	if (m3) goto inst_next;
 	Rt = -1;
 	if (m4) goto inst_next;
+@else
+	Rt = -4;
+	if (m4) goto inst_next;
+	Rt = -3;
+	if (m3) goto inst_next;
+	Rt = -2;
+	if (m2) goto inst_next;
+	Rt = -1;
+	if (m1) goto inst_next;
+@endif
 	Rt = 0;
 }
 
@@ -984,7 +1014,7 @@ macro add_abs_diff(dst, src1, src2, shift)
 
 ########### 16b ############
 
-define token instr16(16)
+define token instr16(16) endian=big
     opsz        = (15, 15)
     opc4        = (11, 14)
     opc5        = (10, 14)
@@ -1015,6 +1045,7 @@ define token instr16(16)
     imm6u       = (0, 5)
     imm7u       = (0, 6)
     imm8s       = (0, 7) signed
+    imm9u       = (0, 8)
     imm10s      = (0, 9) signed
     xwi37_ls    = (7, 7)
     swid9       = (0, 8)
@@ -1130,10 +1161,10 @@ fp_rel7w: [fp + rel7w] is fp & rel7w { addr:4 = fp + rel7w; export addr; }
 rel8: addr is imm8s [ addr = inst_start + (imm8s << 1); ] { export *:4 addr; }
 
 
-:beqs38 rt3b,rel8 is $(I16) & opc4=0b1010 & rt3b & rel8 { if(a5 == rt3b) goto rel8; }
-:bnes38 rt3b,rel8 is $(I16) & opc4=0b1011 & rt3b & rel8 { if(a5 != rt3b) goto rel8; }
-:beqz38 rt3b,rel8 is $(I16) & opc4=0b1000 & rt3b & rel8 { if(rt3b == 0) goto rel8; }
-:bnez38 rt3b,rel8 is $(I16) & opc4=0b1001 & rt3b & rel8 { if(rt3b != 0) goto rel8; }
+:beqs38 rt3b,rel8 is $(I16) & opc4=0b1010 & rt3b & rel8 { branch_if(a5 == rt3b, rel8); }
+:bnes38 rt3b,rel8 is $(I16) & opc4=0b1011 & rt3b & rel8 { branch_if(a5 != rt3b, rel8); }
+:beqz38 rt3b,rel8 is $(I16) & opc4=0b1000 & rt3b & rel8 { branch_if(rt3b == 0, rel8); }
+:bnez38 rt3b,rel8 is $(I16) & opc4=0b1001 & rt3b & rel8 { branch_if(rt3b != 0, rel8); }
 
 :j8    rel8 is $(I16) & opc7=0b1010101     & rel8 { goto rel8; }
 :jr5   rb5  is $(I16) & opc10=0b1011101000 & rb5  { goto [rb5]; }
@@ -1148,8 +1179,8 @@ rel8: addr is imm8s [ addr = inst_start + (imm8s << 1); ] { export *:4 addr; }
 :slt45   ra4, rb5   is $(I16) & opc6=0b110001 & ra4 & rb5   { ta = zext(ra4 < rb5); }
 :slts45  ra4, rb5   is $(I16) & opc6=0b110000 & ra4 & rb5   { ta = zext(ra4 s< rb5); }
 
-:beqzs8 rel8 is $(I16) & opc7=0b1101000 & rel8 { if(ta == 0) goto rel8; }
-:bnezs8 rel8 is $(I16) & opc7=0b1101001 & rel8 { if(ta != 0) goto rel8; }
+:beqzs8 rel8 is $(I16) & opc7=0b1101000 & rel8 { branch_if(ta == 0, rel8); }
+:bnezs8 rel8 is $(I16) & opc7=0b1101001 & rel8 { branch_if(ta != 0, rel8); }
 
 
 ### Misc Instruction ###
@@ -1179,7 +1210,7 @@ sp_rel7w: [+ rel7w] is rel7w { addr:4 = sp + rel7w; export addr; }
 
 imm6u_: imm8 is imm6u [ imm8 = imm6u << 2; ] { export *[const]:4 imm8; }
 :addri36.sp rt3, imm6u_ is $(I16) & opc6=0b011000 & rt3 & imm6u_ { rt3 = sp + imm6u_; }
-:add5.pc rt5b is $(I16) & opc10=0b1011101101 & rt5b { rt5b = pc + rt5b; }
+:add5.pc rt5b is $(I16) & opc10=0b1011101101 & rt5b { rt5b = inst_start + rt5b; }
 :and33 rt3, ra3 is $(I16) & $(MISC33) & rt3 & ra3 & imm3u=0b110 { rt3 = rt3 & ra3; }
 :neg33 rt3, ra3 is $(I16) & $(MISC33) & rt3 & ra3 & imm3u=0b010 { rt3 = -ra3; }
 :not33 rt3, ra3 is $(I16) & $(MISC33) & rt3 & ra3 & imm3u=0b011 { rt3 = ~ra3; }
@@ -1188,7 +1219,7 @@ imm6u_: imm8 is imm6u [ imm8 = imm6u << 2; ] { export *[const]:4 imm8; }
 
 ### Bit Manipulation Instructions (V3 16-bit) ###
 
-:bmski33 rt3, imm3ub is $(I16) & opc6=0b001011 & rt3 & imm3ub & imm3u=0b110 { rt3 = (rt3 >> imm3ub) & 1; }
+:bmski33 rt3, imm3ub is $(I16) & opc6=0b001011 & rt3 & imm3ub & imm3u=0b110 { rt3 = rt3 & (1 << imm3ub); }
 :fexti33 rt3, imm3ub is $(I16) & opc6=0b001011 & rt3 & imm3ub & imm3u=0b111 { rt3 = rt3 & ((1 << (imm3ub + 1)) - 1); }
 
 ### Misc. Instructions (V3 16-bit) ###
@@ -1217,21 +1248,22 @@ push25_re: re2 is re2 & re2=1 { push25_s2(); }
 push25_re: re2 is re2 & re2=2 { push25_s4(); }
 push25_re: re2 is re2 & re2=3 { push25_s8(); }
 
-:push25 push25_re, imm5u__ is $(I16) & opc8=0b11111000 & re2 & push25_re & imm5u__ {
+push25_set_s2: is re2 & re2=0 {}
+push25_set_s2: is re2 { s2 = inst_start & 0xfffffffc; }
+
+:push25 push25_re, imm5u__ is $(I16) & opc8=0b11111000 & re2 & push25_re & push25_set_s2 & imm5u__ {
 	mult_addr = sp;
 	push25_special();
 	build push25_re;
 	sp = mult_addr - imm5u__;
-	if(re2 < 1) goto <end>;
-		s2 = pc & 0xfffffffc;
-	<end>
+    build push25_set_s2;
 }
 
-macro pop25_special() { LmwOp(fp); LmwOp(gp); LmwOp(lp);   }
-macro pop25_s0() { LmwOp(s0); }
-macro pop25_s2() { pop25_s0(); LmwOp(s1); LmwOp(s2); }
-macro pop25_s4() { pop25_s2(); LmwOp(s3); LmwOp(s4); }
-macro pop25_s8() { pop25_s4(); LmwOp(s5); LmwOp(s6); LmwOp(s7); LmwOp(s8); }
+macro pop25_special() { Lmwai(fp); Lmwai(gp); Lmwai(lp); }
+macro pop25_s0() { Lmwai(s0); }
+macro pop25_s2() { pop25_s0(); Lmwai(s1); Lmwai(s2); }
+macro pop25_s4() { pop25_s2(); Lmwai(s3); Lmwai(s4); }
+macro pop25_s8() { pop25_s4(); Lmwai(s5); Lmwai(s6); Lmwai(s7); Lmwai(s8); }
 
 pop25_re: re2 is re2 & re2=0 { pop25_s0(); }
 pop25_re: re2 is re2 & re2=1 { pop25_s2(); }
@@ -1239,10 +1271,10 @@ pop25_re: re2 is re2 & re2=2 { pop25_s4(); }
 pop25_re: re2 is re2 & re2=3 { pop25_s8(); }
 
 :pop25 pop25_re, imm5u__ is $(I16) & opc8=0b11111001 & re2 & pop25_re & imm5u__ {
-	mult_addr = sp;
+	mult_addr = sp + imm5u__;
 	build pop25_re;
 	pop25_special();
-	sp = mult_addr + imm5u__;
+	sp = mult_addr;
 	return [lp];
 }
 
@@ -1261,6 +1293,26 @@ define pcodeop ex9;
 	ex9(imm5u:4);
 }
 
+##########################
+# DSP/ZOL
+
+:mtlbi   Rel16 is $(I32) & $(BR2) & Rt=0b00001 & Br2t=0b0000 & Rel16 { lb = inst_start + Rel16; }
+:mtlei   Rel16 is $(I32) & $(BR2) & Rt=0b00010 & Br2t=0b0000 & Rel16 { le = inst_start + Rel16; }
+
+
+##########################
+# IFC (Inline Function Call) Extension
+
+:ifcall Rel16 is $(I32) & $(BR2) & Rt=0b00000 & Br2t=0b0000 & Rel16 unimpl
+
+:ifret is $(I32) & $(JREG) & Rt=0 & Ra=0 & Rb=0 & DtIt=0b00 & Jz=0b01 & JrHint=1 & Sub5=0b00000 unimpl
+
+rel9: addr is imm9u [ addr = inst_start + (imm9u << 1); ] { export *:4 addr; }
+:ifcall9 rel9 is $(I16) & opc6=0b111100 & rel9 unimpl
+
+# Special case of MOV55 encoding (Rt=sp, Ra=sp)
+:ifret16 is $(I16) & opc5=0b00000 & rt5=0b11111 & ra5=0b11111 unimpl
+
 
 ##########################
 # Floating Point Extension
@@ -1288,12 +1340,12 @@ define pcodeop fcpyss;
 
 define pcodeop fmadds;
 :fmadds Fst, Fsa, Fsb is $(I32) & $(COP) & fop4=0x4 & Fst & Fsa & Fsb  & cop4=0x0 {
-	Fst = fmadds(Fsa, Fsb);
+	Fst = fmadds(Fst, Fsa, Fsb);
 }
 
 define pcodeop fmsubs;
 :fmsubs Fst, Fsa, Fsb is $(I32) & $(COP) & fop4=0x5 & Fst & Fsa & Fsb  & cop4=0x0 {
-	Fst = fmsubs(Fsa, Fsb);
+	Fst = fmsubs(Fst, Fsa, Fsb);
 }
 
 define pcodeop fcmovns;
@@ -1308,12 +1360,12 @@ define pcodeop fcmovzs;
 
 define pcodeop fnmadds;
 :fnmadds Fst, Fsa, Fsb is $(I32) & $(COP) & fop4=0x8 & Fst & Fsa & Fsb  & cop4=0x0 {
-	Fst = fnmadds(Fsa, Fsb);
+	Fst = fnmadds(Fst, Fsa, Fsb);
 }
 
 define pcodeop fnmsubs;
 :fnmsubs Fst, Fsa, Fsb is $(I32) & $(COP) & fop4=0x9 & Fst & Fsa & Fsb  & cop4=0x0 {
-	Fst = fnmsubs(Fsa, Fsb);
+	Fst = fnmsubs(Fst, Fsa, Fsb);
 }
 
 define pcodeop fmuls;
@@ -1410,12 +1462,12 @@ define pcodeop fcpysd;
 
 define pcodeop fmaddd;
 :fmaddd Fdt, Fda, Fdb is $(I32) & $(COP) & fop4=0x4 & Fdt & Fda & Fdb  & cop4=0x8 {
-	Fdt = fmaddd(Fda, Fdb);
+	Fdt = fmaddd(Fdt, Fda, Fdb);
 }
 
 define pcodeop fmsubd;
 :fmsubd Fdt, Fda, Fdb is $(I32) & $(COP) & fop4=0x5 & Fdt & Fda & Fdb  & cop4=0x8 {
-	Fdt = fmsubd(Fda, Fdb);
+	Fdt = fmsubd(Fdt, Fda, Fdb);
 }
 
 define pcodeop fcmovnd;
@@ -1430,12 +1482,12 @@ define pcodeop fcmovzd;
 
 define pcodeop fnmaddd;
 :fnmaddd Fdt, Fda, Fdb is $(I32) & $(COP) & fop4=0x8 & Fdt & Fda & Fdb  & cop4=0x8 {
-	Fdt = fnmaddd(Fda, Fdb);
+	Fdt = fnmaddd(Fdt, Fda, Fdb);
 }
 
 define pcodeop fnmsubd;
 :fnmsubd Fdt, Fda, Fdb is $(I32) & $(COP) & fop4=0x9 & Fdt & Fda & Fdb  & cop4=0x8 {
-	Fdt = fnmsubd(Fda, Fdb);
+	Fdt = fnmsubd(Fdt, Fda, Fdb);
 }
 
 define pcodeop fmuld;
@@ -1481,7 +1533,7 @@ define pcodeop fd2ui;
 
 define pcodeop fd2ui.z;
 :fd2ui.z Fst, Fda is $(I32) & $(COP) & fop4=0xf & Fst & Fda & f2op=0x14 & cop4=0x8 {
-	Fst = fs2ui.z(Fda);
+	Fst = fd2ui.z(Fda);
 }
 
 define pcodeop fd2si;
@@ -1491,7 +1543,7 @@ define pcodeop fd2si;
 
 define pcodeop fd2si.z;
 :fd2si.z Fst, Fda is $(I32) & $(COP) & fop4=0xf & Fst & Fda & f2op=0x1c & cop4=0x8 {
-	Fst = fs2si.z(Fda);
+	Fst = fd2si.z(Fda);
 }
 
 # FPU_FS2
@@ -1619,7 +1671,7 @@ OffImm12s: (offs) is Imm12s [ offs = Imm12s << 2; ] { export *[const]:4 offs; }
 	Rt = fpcsr;
 }
 
-:fmtcsr Rt is $(I32) & $(COP) & fop4=0xc & Rt & Fsa & f2op=0x1 & cop4=0x9 {
+:fmtcsr Rt is $(I32) & $(COP) & fop4=0xc & Rt & Fsa=0 & f2op=0x1 & cop4=0x9 {
 	fpcsr = Rt;
 }
 
-- 
2.45.1

